File: Python/Sudoku Solver.py
class Solution:
    # @param board, a 9x9 2D array
    # Solve the Sudoku by modifying the input board in-place.
    # Do not return any value.
    def solveSudoku(self, board):
        self.board = board
        self.fillBoard()
            
    def isValid(self, row, col, num):
        for i in range(9):
            if self.board[i][col] == num or self.board[row][i] == num:
                    return False
        for i in range(row-(row%3), row-(row%3) + 3):
            for j in range(col - (col%3), col - (col%3) + 3):
                if self.board[i][j] == num:
                    return False
        return True
        
    def isEmpty(self):
        for i in range(9):
            for j in range(9):
                if self.board[i][j] == ".":
                    return i, j
        return -1, -1
        
    def fillBoard(self):
        i,j = self.isEmpty()
        if i == -1 and j == -1:
            return True
        for num in  ["1","2","3","4","5","6","7","8","9"]:
            if self.isValid(i,j,num):
                self.board[i][j] = num
                if self.fillBoard():
                    return True
                self.board[i][j] = "."
        return False
        
        

File: Java/src/Sudoku Solver.java
public class Solution
{
  private char[][] board;

  public void solveSudoku(char[][] board)
  {
    this.board = board;
    fillBoard();
  }

  private boolean isValid(int row, int col, char num)
  {
    for (int i = 0; i < 9; i++)
    {
      if (board[i][col] == num || board[row][i] == num)
        return false;
    }

    for (int i = row - (row % 3); i < row - (row % 3) + 3; i++)
    {
      for (int j = col - (col % 3); j < col - (col % 3) + 3; j++)
      {

        if (board[i][j] == num)
          return false;
      }
    }
    return true;
  }

  private int[] isEmpty()
  {
    int[] ret = new int[2];
    for (int row = 0; row < 9; row++)
    {
      for (int col = 0; col < 9; col++)
      {
        if (board[row][col] == '.')
        {
          ret[0] = row;
          ret[1] = col;
          return ret;
        }
      }
    }
    ret[0] = -1;
    ret[1] = -1;
    return ret;
  }

  private boolean fillBoard()
  {
    int[] emptyPoint = new int[2];
    emptyPoint = isEmpty();
    if (emptyPoint[0] == -1 && emptyPoint[1] == -1)
      return true;
    int row = emptyPoint[0];
    int col = emptyPoint[1];

    for (char num = '1'; num <= '9'; num++)
    {
      if (isValid(row, col, num))
      {
        board[row][col] = num;
        if (fillBoard())
        {
          return true;
        }
        board[row][col] = '.';
      }
    }
    return false;
  }
}
